doc/how-to-source-packages.rst

```diff
@@ -0,0 +1,309 @@
+
+Fields pointing to *source code* repositories
+---------------------------------------------
+
+Source [code] repositories are a way to specify where to find the source code
+for a package, both for :ref:`package authors<pkg-authors>` and maintainers and
+for :ref:`package consumers<pkg-consumers>`.
+
+A relatively structured set of version control system (VCS) fields, that vary
+depending on the :ref:`VCS kind<vcs-kind>`, enables Cabal commands and other
+tools to interpret and make effective use of this information.
+
+``source-repository``
+   Says where to find the source for a package and is a metadata field of a
+   **package description**. This is the source marker.
+
+``source-repository-package``
+   Locates the source for one or more **project dependencies** that live in the
+   same **source code repository** but in separate folders. The root of the
+   repository is assumed when the ``subdir`` field is omitted.
+
+.. Warning::
+
+   The ``source-repository`` field, the one without "package" is for a package!
+   This actually makes perfect sense as it's within the package description
+   looking outwards.
+
+   .. code-block:: text
+
+       (package description)      source-repository         (Git repository)
+      solver-benchmarks.cabal ──────────────────────────⮚ github/haskell/cabal
+
+      $ cat solver-benchmarks.cabal
+      ...
+      source-repository head
+         type:     git
+         location: https://github.com/haskell/cabal/
+         subdir:   solver-benchmarks
+      ...
+      library
+         build-depends:
+      ...
+            vector
+      ...
+
+   The ``source-repository-package`` field, the one with "package" isn't for a
+   package! It's for a project dependency. This name is a bit misleading because
+   it can describe multiple dependencies and not just one. That it is for
+   "package" also makes sense. Ultimately it is the union of all dependencies of
+   local packages within the project that actually declare the set of project
+   dependencies, not the project.  The project can tighten version constraint
+   ranges with ``constraints`` or loosen them with ``allow-newer`` or
+   ``allow-older`` but it cannot add or remove dependencies.
+
+   This project stanza is like a redirect or override for dependencies of all
+   the packages in the project saying "use this source code repository" for this
+   package dependency, not Hackage.
+
+   .. code-block:: text
+
+        (project)       source-repository-package         (Git repository)
+      cabal.project ──────────────────────────────────⮚ github/haskell/vector
+
+      $ tree
+      ...
+      └── solver-benchmarks
+         └── solver-benchmarks.cabal
+
+      $ cat cabal.project
+      packages: solver-benchmarks
+      ...
+      source-repository-package
+         type:     git
+         location: https://github.com/haskell/vector.git
+         subdir:   vector
+
+.. _pkg-consumers:
+
+*Source code* when taking a package dependency
```

Comment:
```suggestion
Taking *Source code* as a package dependency
```

Reply:
I disagree with this change of section title. The section talks about how Cabal deals with source code as well. The section starts with this before it goes on to talk about how to deal with dependency source code ourselves.

Reply:
@malteneuss if you agree, can you please resolve this conversation?

Reply:
I can't due to https://github.com/orgs/community/discussions/8327 xD

Reply:
OK, I can mark it as resolved.